### DISCUSSION TITLE
Compiling project in docker

### DISCUSSION THREAD

**Discussion by ezamosch** [microsoft/vcpkg#37628] [Q&A]
Hello!

I am trying to compile c++ project in docker using vcpkg. What I have atm:

**PROJECT TREE:**
```
 -- docker/
     -- Dockerfile
 -- src/
     -- index.cpp
     -- vcpkg.json
```

**FILES:**
**vcpkg.json**
```
{
  "dependencies": [
    "cpr"
  ]
}
```

**index.cpp**
```
#include <iostream>
#include <cpr/cpr.h>

int main(){
    std::cout << "hello" << "\n"; 
}
```

**Dockerfile**
```
FROM gcc:latest

COPY /src/. /usr/src/test
WORKDIR /usr/src/test

RUN apt-get update
RUN apt-get install -y tar zip unzip 

RUN git clone https://github.com/microsoft/vcpkg
RUN ./vcpkg/bootstrap-vcpkg.sh
RUN ./vcpkg/vcpkg install

RUN g++ -o Invest-app index.cpp
CMD [ "./Invest-app" ]
```

Now when I build image I get error on "RUN g++ -o Invest-app index.cpp" command:
```
 > [9/9] RUN g++ -o Invest-app index.cpp:
0.462 index.cpp:5:10: fatal error: cpr/cpr.h: No such file or directory
0.462     5 | #include <cpr/cpr.h>
0.462       |          ^~~~~~~~~~~
0.462 compilation terminated.
```

So basically I want most convenient and easy way to link all dependencies in docker (including copying all necessary files in work dir like dll etc). I am a huge fan of vcpkg in windows, because this tool abstract out all headache with linking and by simple command `.\vcpkg\vcpkg integrate install` I forget about linking theme at all.

So can somebody help to compile all this? Maybe there is some kind of command for linux os that will associate all necessary dependencies in my project after I do `RUN ./vcpkg/vcpkg install`?

**Comment by dg0yt** [comment]
~~~
RUN g++ -o Invest-app index.cpp
~~~
How is this expected to know the include dirs and link libaries?

**Reply by ezamosch** [reply]
Nohow. That is why I asked for help here.

**Comment by ezamosch** [comment]
I started my program by linking the libs:
```
FROM gcc:latest

COPY /src/. /usr/src/test

WORKDIR /usr/src/test

RUN apt-get update

RUN apt-get install -y tar zip unzip 
#RUN apt-get install -y curl build-essential

RUN git clone https://github.com/microsoft/vcpkg
RUN ./vcpkg/bootstrap-vcpkg.sh
RUN ./vcpkg/vcpkg install
RUN g++ -o Invest-app -I/usr/src/test/vcpkg_installed/x64-linux/include -L/usr/src/test/vcpkg_installed/x64-linux/lib index.cpp

CMD [ "./Invest-app" ]
```

But the question for me is still actual - is there any elegant solution to not include all needed libraries manually (for example as I mentioned vcpkg integrate install for VS).

**Reply by dg0yt** [reply]
If you really want to use libs, `-L...`  won't be enough.

> is there any elegant solution to not include all needed libraries manually (for example as I mentioned vcpkg integrate install for VS).

The msbuild integration adds one include dir (not enough) and all libs (too much). Not so elegant for non-trivial setups.
In general, the elegant solutions start with pkgconfig and continue with CMake.

**Reply by ezamosch** [reply]
@dg0yt can you please provide some examples on this lil project? how you would organize it?

**Comment by ezamosch** [comment]
Compiled it manually linking everything it needs (Docker RUN cmd):
`RUN g++ -std=c++17 index.cpp -o Cpp-service -Iinclude -Ivcpkg_installed/x64-linux/include -Lvcpkg_installed/x64-linux/lib -lcpr -lcurl -lz -lssl -lcrypto`
